sw/host/opentitanlib/src/spiflash/flash.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::io::spi::{Target, Transfer};
+use crate::spiflash::sfdp::{BlockEraseSize, Sfdp, SupportedAddressModes};
+use anyhow::{ensure, Result};
+use std::convert::TryFrom;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("address out of bounds: {0} >= {1}")]
+    AddressOutOfBounds(u32, u32),
+    #[error("erase at address {0} not aligned on {1}-byte boundary")]
+    BadEraseAddress(u32, u32),
+    #[error("erase length {0} not a multiple of {1} bytes")]
+    BadEraseLength(u32, u32),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum AddressMode {
+    Mode3b,
+    Mode4b,
+}
+
+impl From<SupportedAddressModes> for AddressMode {
+    fn from(mode: SupportedAddressModes) -> Self {
+        match mode {
+            SupportedAddressModes::Mode4b => AddressMode::Mode4b,
+            _ => AddressMode::Mode3b,
+        }
+    }
+}
+
+impl Default for AddressMode {
+    fn default() -> Self {
+        AddressMode::Mode3b
+    }
+}
+
+pub struct SpiFlash {
+    pub size: u32,
+    pub erase_size: u32,
+    pub erase_opcode: u8,
+    pub program_size: u32,
+    pub address_mode: AddressMode,
+    pub sfdp: Option<Sfdp>,
+}
+
+impl Default for SpiFlash {
+    fn default() -> Self {
+        SpiFlash {
+            size: 16 * 1024 * 1024,
+            erase_size: 4 * 1024,
+            erase_opcode: 0x20,
+            program_size: SpiFlash::LEGACY_PAGE_SIZE,
+            address_mode: AddressMode::default(),
+            sfdp: None,
+        }
+    }
+}
+
+impl SpiFlash {
+    // Well known SPI Flash opcodes.
+    pub const READ: u8 = 0x03;
+    pub const PAGE_PROGRAM: u8 = 0x02;
+    pub const SECTOR_ERASE: u8 = 0x20;
+    pub const CHIP_ERASE: u8 = 0x60;
+    pub const WRITE_ENABLE: u8 = 0x06;
+    pub const WRITE_DISABLE: u8 = 0x04;
+    pub const READ_STATUS: u8 = 0x05;
+    pub const READ_ID: u8 = 0x9f;
+    pub const ENTER_4B: u8 = 0xb7;
+    pub const EXIT_4B: u8 = 0xe9;
+    pub const READ_SFDP: u8 = 0x5a;
+    pub const NOP: u8 = 0x00;
+    pub const RESET_ENABLE: u8 = 0x66;
+    pub const RESET: u8 = 0x99;
+
+    // The legacy JEDEC page size for programming operations is 256 bytes.
+    pub const LEGACY_PAGE_SIZE: u32 = 256;
+
+    // Status register bits:
+    // The `WIP` bit indicates a write in progress (sometimes called the busy bit).
+    pub const STATUS_WIP: u8 = 0x01;
+    // The `WEL` bit is the write enable latch.
+    pub const STATUS_WEL: u8 = 0x02;
+
+    // Prepare an opcode and address buffer according to the address_mode.
```

Comment:
These _should_ be doc comments. Hehe.

Reply:
meh.  Why then should "well known spi flash opcodes" not be a doc comment?  Those constants are used within `SpiFlash`, but could also be used by an end-user to construct her own SPI transactions.